README.md
# KUparser.jl
Dependency parsing with word vectors.


ChangeLog
2017-05-08  Deniz Yuret  <dyuret@ku.edu.tr>

	* TODO:
	+ dropout: run experiments
	+ arctype
	+ parser features
	+ embedding dimensions
	+ hyperopt: hidden size
	- use lcword, lemma, xpos, feats, deps, misc as features
	- finetuning: try finetuning word vectors
	- dynamic training vs static training
	- ensemble
	- separate move and label pred
	- Should try a composition function instead of ABabL features.
	- Should try LSTM instead of MLP to make decisions, or find a way to represent stack, attention?
	- beam: figure out why beam training is not helping, compare with old bparser, larger beam, different loss, no earlystop, global normalization? also why slow down?
	- cnn: try cnn word vectors

	* arctype-experiments: no improvement
	(:arctype,"ArcHybridR1")(:feats,"FEATS")	0.756962690488702
	(:arctype,"ArcHybrid13")(:feats,"Flist.hybrid25")	0.6942546855841654
	(:arctype,"ArcEager13")(:feats,"Flist.eager25")	0.7165002627430374
	(:arctype,"ArcEager13")(:feats,"Flist.zn11cpv")	0.7413732702750043
	(:arctype,"ArcEagerR1")(:feats,"Flist.zn11cpv")	0.7355929234541951
	(:arctype,"ArcHybridR1")(:feats,"Flist.gn13cpv")	0.7363811525661237
	(:arctype,"ArcEagerR1")(:feats,"Flist.gn13cpv")	0.7517954107549484
	(:arctype,"ArcEager13")(:feats,"Flist.eager27")	0.7196531791907514
	(:arctype,"ArcHybrid13")(:feats,"Flist.hybrid25b")	0.6914520931861973
	(:arctype,"ArcEager13")(:feats,"Flist.eager20")	0.7130846032580137
	(:arctype,"ArcEager13")(:feats,"Flist.eager23")	0.7078297425118234
	(:arctype,"ArcEager13")(:feats,"Flist.eager21")	0.7133473462953231
	(:arctype,"ArcHybrid13")(:feats,"Flist.hybrid28")	0.7053774741636013

	* feature-experiments: no improvement.
	[ec2-user@ip-172-31-7-57 features]$ ./mktable.pl features1.log (without dropout)
	mlpdims=(4224,4096,73) 	0.7520581537922578
	mlpdims=(7125,4096,73) 	0.7431248905237344
	mlpdims=(5191,4096,73) 	0.742862147486425
	mlpdims=(5322,4096,73) 	0.7506568575932738
	[ec2-user@ip-172-31-7-57 features]$ ./mktable.pl features2.log (with dropout)
	mlpdims=(4224,4096,73) 	0.765458048695043
	mlpdims=(7125,4096,73) 	0.7537221930285514
	mlpdims=(5191,4096,73) 	0.7640567524960589
	mlpdims=(5322,4096,73) 	0.765808372744789

	* dropout-experiments: 0.5,0.5 seems reasonable, brings 1.8%
	(:dropout,[0.0,0.0])	0.7520581537922578
	(:dropout,[0.0,0.25])	0.7551234892275355
	(:dropout,[0.0,0.5])	0.757313014538448
	(:dropout,[0.0,0.75])	0.7532842879663689
	(:dropout,[0.25,0.0])	0.7618672271851462
	(:dropout,[0.25,0.25])	0.7610789980732178
	(:dropout,[0.25,0.5])	0.7609914170607812
	(:dropout,[0.25,0.75])	0.7587143107374321
	(:dropout,[0.5,0.0])	0.7678227360308285
	(:dropout,[0.5,0.25])	0.7722893676650902
	(:dropout,[0.5,0.5])	0.7699246803293046
	(:dropout,[0.5,0.75])	0.7679978980557015
	(:dropout,[0.75,0.0])	0.7480294272201786
	(:dropout,[0.75,0.25])	0.7547731651777895
	(:dropout,[0.75,0.5])	0.7302504816955684
	(:dropout,[0.75,0.75])	0.7317393589069889

	* embedsize-experiments: (using 0.5 dropout) brings another 1.8%
	(17,37,10),	# mlpdims=(4224,4096,73) 	0.7644946575582414
	(128,128,16),   # mlpdims=(5432,4096,73) 	0.7826239271325977

	(32,37,10),     # mlpdims=(4284,4096,73) 	0.7714135575407252
	(64,37,10),     # mlpdims=(4412,4096,73) 	0.7772814853739709
	(128,37,10),    # mlpdims=(4668,4096,73) 	0.7813102119460501
	(256,37,10),    # mlpdims=(5180,4096,73) 	0.7791206866351374

	(17,32,10),     # mlpdims=(4184,4096,73) 	0.7652828866701699
	(17,64,10),     # mlpdims=(4440,4096,73) 	0.7644070765458049
	(17,128,10),    # mlpdims=(4952,4096,73) 	0.7610789980732178
	(17,256,10),    # mlpdims=(5976,4096,73) 	0.7609914170607812

	(17,37,8),      # mlpdims=(4212,4096,73) 	0.765808372744789 
	(17,37,16),     # mlpdims=(4260,4096,73) 	0.7588018917498687
	(17,37,32),     # mlpdims=(4356,4096,73) 	0.7629181993343843
	(17,37,64),     # mlpdims=(4548,4096,73) 	0.7679103170432651

        (128,16,10), 	# mlpdims=(4500,4096,73) 	0.7757926081625504
        (128,32,10),    # mlpdims=(4628,4096,73) 	0.7806971448589946
        (128,64,10),    # mlpdims=(4884,4096,73) 	0.7803468208092486
        (128,128,10),   # mlpdims=(5396,4096,73) 	0.7771939043615345
        (128,256,10),   # mlpdims=(6420,4096,73) 	0.7789455246102645

        (128,32,8),     # mlpdims=(4616,4096,73) 	0.7771063233490979
***     (128,32,16),    # mlpdims=(4664,4096,73) 	0.7810474689087405
        (128,32,32),    # mlpdims=(4760,4096,73) 	0.778157295498336 
        (128,32,64),    # mlpdims=(4952,4096,73) 	0.7775442284112805
        (128,32,128),   # mlpdims=(5336,4096,73) 	0.77955859169732  

	* hiddensize-experiments:
	no-dropout:
	(:hidden,[1024])	0.7425994044491154
	(:hidden,[1024,1024])	0.743562795585917
	(:hidden,[2048])	0.7461026449465756
	(:hidden,[2048,2048])	0.7538973550534244
	(:hidden,[4096])	0.7525836398668768
	(:hidden,[4096,4096])	0.7550359082150989
	(:hidden,[8192])	0.7549483272026625
	(:hidden,[8192,8192])	0.7597652828866702
	(:hidden,[16384])	0.7589770537747417
	(:hidden,[16384,16384])	out-of-memory
	with-dropout:
	(:hidden,[1024])	0.766333858819408
	(:hidden,[1024,1024])	0.7518829917673848
	(:hidden,[2048])	0.7657207917323524
	(:hidden,[2048,2048])	0.7680854790681381
	(:hidden,[4096])	0.7695743562795586
	(:hidden,[4096,4096])	0.7662462778069714
	(:hidden,[8192])	0.7703625853914872
	(:hidden,[8192,8192])	0.7598528638991067
	(:hidden,[16384])	0.766859344894027
	(:hidden,[16384,16384])	out-of-memory
	with-new-embedsizes: (and 30 epochs)
	(:hidden,[32])  	"oracle30"	0.6986337362059906
	(:hidden,[64])  	"oracle29"	0.7458399019092661
	(:hidden,[128]) 	"oracle29"	0.7714135575407252
	(:hidden,[256]) 	"oracle26"	0.7856016815554387
	(:hidden,[512]) 	"oracle30"	0.7903310562270099
	(:hidden,[1024])	"oracle28"	0.7947976878612717
	(:hidden,[2048])	"oracle27"	0.7935715536871606
	(:hidden,[4096])	"oracle28"	0.7943597827990891
	(:hidden,[8192])	"oracle28"	0.7902434752145735
	(:hidden,[16384])	"oracle25"	0.7838500613067088

	(:hidden,[32,32])	"oracle30"	0.571466106148187
	(:hidden,[64,64])	"oracle26"	0.7153617095813628

src/conll17.jl
# CONLL-U FORMAT:
# 1. ID: Word index, integer starting at 1 for each new sentence; may be a range for multiword tokens; may be a decimal number for empty nodes.
# 2. FORM: Word form or punctuation symbol.
# 3. LEMMA: Lemma or stem of word form.
# 4. UPOSTAG: Universal part-of-speech tag.
# 5. XPOSTAG: Language-specific part-of-speech tag; underscore if not available.
# 6. FEATS: List of morphological features from the universal feature inventory or from a defined language-specific extension; underscore if not available.
# 7. HEAD: Head of the current word, which is either a value of ID or zero (0).
# 8. DEPREL: Universal dependency relation to the HEAD (root iff HEAD = 0) or a defined language-specific subtype of one.
# 9. DEPS: Enhanced dependency graph in the form of a list of head-deprel pairs.
# 10. MISC: Any other annotation.

function readconllu(filename::AbstractString,wordvecs=nothing)
    v = Vocab()
    v.postags = UPOSTAG
    v.deprels = UDEPREL
    words = Dict{String,WordId}()
    lcwords = Dict{String,Vector{WordId}}()
    sentences = Sentence[]
    s = Sentence(); s.vocab = v
    nw = 0
    info(filename)
    for line in eachline(filename)
        line = chomp(line)
        fields = split(line, '\t') # TODO: this makes fields[2] a substring preventing garbage collection.
        if line == ""
            if nw > 0
                nw = 0
                push!(sentences,s)
                s = Sentence(); s.vocab = v
            end
        elseif line[1] == '#'
            continue
        elseif ismatch(r"^\d+$", fields[1])
            nw += 1
            if fields[1] != "$nw"; error("Token out of order in [$line]"); end

            if !haskey(words, fields[2])
                id = 1 + length(words)
                wform = String(fields[2])
                lform = lowercase(wform)
                words[wform] = id
                if !haskey(lcwords, lform); lcwords[lform] = WordId[]; end
                push!(lcwords[lform], id)
                push!(v.words, wform)
            end
            push!(s.word, words[fields[2]])

            hd = parse(Position,fields[7])
            if !(0 <= hd <= typemax(eltype(s.head))); Base.warn_once("Bad head: $hd"); hd=0; end
            push!(s.head, hd)

            pt = findfirst(v.postags, fields[4])
            if pt == 0; Base.warn_once("Bad postag: $(fields[4])"); end
            push!(s.postag, pt)

            dr = fields[8]
            # Get rid of language specific extensions
            drcolon = search(dr, ':')
            if drcolon > 0; dr = dr[1:drcolon-1]; end
            dr = findfirst(v.deprels, dr)
            if dr == 0; Base.warn_once("Bad deprel: $(fields[8])"); end
            push!(s.deprel, dr)

        elseif ismatch(r"^\d+[-.]\d+$", fields[1])
            # skip
        else
            error("Cannot parse [$line]")
        end
    end
    if wordvecs != nothing
        info(wordvecs)
        open(`xzcat $wordvecs`) do io
            firstline = split(readline(io)) # first line gives count and dim
            wdims = parse(Int,firstline[2])
            # TODO: We are leaving unknown word vectors as zero!
            v.wvecs = zeros(WVtype, wdims, length(v.words))
            for line in eachline(io)
                n = search(line, ' ')
                w = SubString(line,1,n-1)
                if !haskey(lcwords,w); continue; end
                a = map(s->parse(Float32,s), split(SubString(line,n+1)))
                if wdims != length(a); throw(DimensionMismatch()); end
                for i in lcwords[w]
                    v.wvecs[:,i] = a
                end
            end
        end
    end
    return sentences
end

# Universal POS tags (17)
const UPOSTAG = String[ # Dict{String,UInt8}(
"ADJ"   ,# => 1, # adjective
"ADP"   ,# => 2, # adposition
"ADV"   ,# => 3, # adverb
"AUX"   ,# => 4, # auxiliary
"CCONJ" ,# => 5, # coordinating conjunction
"DET"   ,# => 6, # determiner
"INTJ"  ,# => 7, # interjection
"NOUN"  ,# => 8, # noun
"NUM"   ,# => 9, # numeral
"PART"  ,# => 10, # particle
"PRON"  ,# => 11, # pronoun
"PROPN" ,# => 12, # proper noun
"PUNCT" ,# => 13, # punctuation
"SCONJ" ,# => 14, # subordinating conjunction
"SYM"   ,# => 15, # symbol
"VERB"  ,# => 16, # verb
"X"     ,# => 17, # other
]

# Universal Dependency Relations (37)
const UDEPREL = String[ # Dict{String,UInt8}(
"root"       ,# => 1, # root
"acl"        ,# => 2, # clausal modifier of noun (adjectival clause)
"advcl"      ,# => 3, # adverbial clause modifier
"advmod"     ,# => 4, # adverbial modifier
"amod"       ,# => 5, # adjectival modifier
"appos"      ,# => 6, # appositional modifier
"aux"        ,# => 7, # auxiliary
"case"       ,# => 8, # case marking
"cc"         ,# => 9, # coordinating conjunction
"ccomp"      ,# => 10,# clausal complement
"clf"        ,# => 11, # classifier
"compound"   ,# => 12, # compound
"conj"       ,# => 13, # conjunct
"cop"        ,# => 14, # copula
"csubj"      ,# => 15, # clausal subject
"dep"        ,# => 16, # unspecified dependency
"det"        ,# => 17, # determiner
"discourse"  ,# => 18, # discourse element
"dislocated" ,# => 19, # dislocated elements
"expl"       ,# => 20, # expletive
"fixed"      ,# => 21, # fixed multiword expression
"flat"       ,# => 22, # flat multiword expression
"goeswith"   ,# => 23, # goes with
"iobj"       ,# => 24, # indirect object
"list"       ,# => 25, # list
"mark"       ,# => 26, # marker
"nmod"       ,# => 27, # nominal modifier
"nsubj"      ,# => 28, # nominal subject
"nummod"     ,# => 29, # numeric modifier
"obj"        ,# => 30, # object
"obl"        ,# => 31, # oblique nominal
"orphan"     ,# => 32, # orphan
"parataxis"  ,# => 33, # parataxis
"punct"      ,# => 34, # punctuation
"reparandum" ,# => 35, # overridden disfluency
"vocative"   ,# => 36, # vocative
"xcomp"      ,# => 37, # open clausal complement
]

# Universal Features (21) with number of possible values
# Abbr: abbreviation (1) Yes (all features also have None as an extra choice)
# Animacy: animacy (4) Anim,Hum,Inan,Nhum
# Aspect: aspect (6) Hab,Imp,Iter,Perf,Prog,Prosp
# Case: case (32)
# Definite: definiteness or state (5) Com,Cons,Def,Ind,Spec
# Degree: degree of comparison (5) Abs,Cmp,Equ,Pos,Sup
# Evident: evidentiality (2) Fh,Nfh
# Foreign: is this a foreign word? (1) Yes
# Gender: gender (4) Com,Fem,Masc,Neut
# Mood: mood (12) Adm,Cnd,Des,Imp,Ind,Jus,Nec,Opt,Pot,Prp,Qot,Sub
# NumType: numeral type (7) Card,Dist,Frac,Mult,Ord,Range,Sets
# Number: number (10) CollCountDualGrpaGrplInvPaucPlurPtanSingTri
# Person: person (5) 0,1,2,3,4
# Polarity: polarity (2) Neg,Pos
# Polite: politeness (4) ElevFormHumbInfm
# Poss: possessive (1) Yes
# PronType: pronominal type (11) ArtDemEmpExcIndIntNegPrsRcpRelTot
# Reflex: reflexive (1) Yes
# Tense: tense (5) FutImpPastPqpPres
# VerbForm: form of verb or deverbative (8) ConvFinGdvGerInfPartSupVnoun
# Voice: voice (8) ActAntipCauDirInvMidPassRcp

# conll17 word vectors
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Ancient_Greek/grc.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Arabic/ar.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Basque/eu.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Bulgarian/bg.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Catalan/ca.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/ChineseT/zh.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Croatian/hr.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Czech/cs.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Danish/da.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Dutch/nl.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/English/en.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Estonian/et.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Finnish/fi.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/French/fr.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Galician/gl.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/German/de.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Greek/el.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Hebrew/he.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Hindi/hi.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Hungarian/hu.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Indonesian/id.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Irish/ga.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Italian/it.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Japanese/ja.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Kazakh/kk.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Korean/ko.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Latin/la.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Latvian/lv.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Norwegian-Bokmaal/no_bokmaal.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Norwegian-Nynorsk/no_nynorsk.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Old_Church_Slavonic/cu.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Persian/fa.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Polish/pl.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Portuguese/pt.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Romanian/ro.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Russian/ru.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Slovak/sk.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Slovenian/sl.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Spanish/es.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Swedish/sv.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Turkish/tr.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Ukrainian/uk.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Urdu/ur.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Uyghur/ug.vectors.xz
# /mnt/ai/data/nlp/conll17/word-embeddings-conll17/Vietnamese/vi.vectors.xz

nothing
